File: ch9/py/src/ingestion_graph/configuration.py
"""Define the configurable parameters for the index graph."""

from __future__ import annotations

from dataclasses import dataclass, field, fields
from typing import Annotated, Literal, Optional, Type, TypeVar, Any
from langchain_core.runnables import RunnableConfig, ensure_config


DEFAULT_DOCS_FILE = "src/docSplits.json"


@dataclass(kw_only=True)
class IndexConfiguration:
    """Configuration class for indexing and retrieval operations.

    This class defines the parameters needed for configuring the indexing and
    retrieval processes, including embedding model selection, retriever provider choice, and search parameters.
    """

    docs_file: str = field(
        default=DEFAULT_DOCS_FILE,
        metadata={
            "description": "Path to a JSON file containing default documents to index."
        },
    )

    embedding_model: Annotated[
        str,
        {"__template_metadata__": {"kind": "embeddings"}},
    ] = field(
        default="openai/text-embedding-3-small",
        metadata={
            "description": "Name of the embedding model to use. Must be a valid embedding model name."
        },
    )

    retriever_provider: Annotated[
        Literal["supabase", "chroma"],
        {"__template_metadata__": {"kind": "retriever"}},
    ] = field(
        default="chroma",
        metadata={
            "description": "The vector store provider to use for retrieval. Options are 'supabase', or 'chroma'."
        },
    )

    search_kwargs: dict[str, Any] = field(
        default_factory=dict,
        metadata={
            "description": "Additional keyword arguments to pass to the search function of the retriever."
        },
    )

    @classmethod
    def from_runnable_config(
        cls: Type[T], config: Optional[RunnableConfig] = None
    ) -> T:
        """Create an IndexConfiguration instance from a RunnableConfig object.

        Args:
            cls (Type[T]): The class itself.
            config (Optional[RunnableConfig]): The configuration object to use.

        Returns:
            T: An instance of IndexConfiguration with the specified configuration.
        """
        config = ensure_config(config)
        configurable = config.get("configurable") or {}
        _fields = {f.name for f in fields(cls) if f.init}
        return cls(**{k: v for k, v in configurable.items() if k in _fields})


T = TypeVar("T", bound=IndexConfiguration)

File: ch9/js/src/ingestion_graph/configuration.ts
import { Annotation } from '@langchain/langgraph';
import { RunnableConfig } from '@langchain/core/runnables';

// This path points to the directory containing the documents to index.
const DEFAULT_DOCS_PATH = 'src/sample_docs.json';

/**
 * The configuration for the indexing process.
 */
export const IndexConfigurationAnnotation = Annotation.Root({
  /**
   * Path to folder containing default documents to index.
   */
  docsPath: Annotation<string>,

  /**
   * Name of the openai embedding model to use. Must be a valid embedding model name.
   */
  embeddingModel: Annotation<'text-embedding-3-small'>,

  /**
   * The vector store provider to store the embeddings.
   * Options are 'supabase', 'chroma'.
   */
  retrieverProvider: Annotation<'supabase' | 'chroma'>,

  /**
   * Whether to index sample documents specified in the docsPath.
   */
  useSampleDocs: Annotation<boolean>,
});

/**
 * Create an typeof IndexConfigurationAnnotation.State instance from a RunnableConfig object.
 *
 * @param config - The configuration object to use.
 * @returns An instance of typeof IndexConfigurationAnnotation.State with the specified configuration.
 */
export function ensureIndexConfiguration(
  config: RunnableConfig
): typeof IndexConfigurationAnnotation.State {
  const configurable = (config?.configurable || {}) as Partial<
    typeof IndexConfigurationAnnotation.State
  >;
  return {
    docsPath: configurable.docsPath || DEFAULT_DOCS_PATH,
    embeddingModel: configurable.embeddingModel || 'text-embedding-3-small',
    retrieverProvider: configurable.retrieverProvider || 'supabase',
    useSampleDocs: configurable.useSampleDocs || false,
  };
}
